File: backend/internal/api/__tests__/onboarding_probe_models_test.go
package api_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/MIEnchating/sub2api-console/backend/internal/api"
	"github.com/MIEnchating/sub2api-console/backend/internal/config"
	"github.com/MIEnchating/sub2api-console/backend/internal/onboarding"
	"github.com/MIEnchating/sub2api-console/backend/internal/taskstore"
)

type probeModelsOnboarding struct {
	api.OnboardingService
	requests []onboarding.Request
}

func (service *probeModelsOnboarding) Enqueue(_ context.Context, request onboarding.Request) (taskstore.Task, error) {
	service.requests = append(service.requests, request)
	return taskstore.Task{ID: "isolated-onboarding", Status: "queued"}, nil
}

func TestOnboardingAPIValidatesProbeModelsBeforeEnqueue(t *testing.T) {
	for _, scenario := range []struct {
		name   string
		models any
		status int
		count  int
	}{
		{"selected", []string{" model-a ", "model-b", "model-a"}, 200, 2},
		{"empty fallback", []string{}, 200, 0},
		{"null rejected", nil, 422, 0},
		{"wrong type", "model-a", 422, 0},
		{"empty model", []string{" "}, 422, 0},
		{"long model", []string{strings.Repeat("x", 257)}, 422, 0},
	} {
		t.Run(scenario.name, func(t *testing.T) {
			service := &probeModelsOnboarding{}
			router := api.New(config.Config{AdminToken: "isolated-token"}, nil, nil, api.Dependencies{Onboarding: service})
			body, err := json.Marshal(map[string]any{"host": "upstream.example", "upstream_type": "sub2api", "upstream_group_id": "7", "local_group_ids": []int{3}, "test_models": scenario.models})
			if err != nil {
				t.Fatal(err)
			}
			request := httptest.NewRequest(http.MethodPost, "/api/onboarding", strings.NewReader(string(body)))
			request.Header.Set("Authorization", "Bearer isolated-token")
			request.Header.Set("Content-Type", "application/json")
			response := httptest.NewRecorder()
			router.ServeHTTP(response, request)
			if response.Code != scenario.status {
				t.Fatalf("status=%d body=%s", response.Code, response.Body.String())
			}
			if scenario.status == 200 && (len(service.requests) != 1 || len(service.requests[0].TestModels) != scenario.count) {
				t.Fatalf("probe models lost in API adaptation: %+v", service.requests)
			}
			if scenario.status != 200 && len(service.requests) != 0 {
				t.Fatal("invalid models reached background work")
			}
		})
	}
}

File: frontend/src/features/upstreams/lib/__tests__/onboarding-probe-models.test.ts
import { expect, it } from "vitest";
import type { OnboardingRequest } from "@/api";
import { onboardingAccountPreviewID, withOnboardingProbeModels } from "../onboarding-probe-models";

const created: OnboardingRequest = {
  host: "upstream.example",
  upstream_type: "sub2api",
  upstream_group_id: "7",
  local_group_ids: [3],
};

it("两个新增账号分别读取自身模型，即使请求顺序调整也不串号", () => {
  const second = { ...created, local_group_ids: [4] };
  const existing = { ...created, account_ids: ["41"] };
  const models = {
    [onboardingAccountPreviewID(created)]: ["gpt-5.2"],
    [onboardingAccountPreviewID(second)]: ["gpt-5.1"],
  };
  expect(withOnboardingProbeModels([second, existing, created], models)).toEqual([
    { ...second, test_models: ["gpt-5.1"] },
    existing,
    { ...created, test_models: ["gpt-5.2"] },
  ]);
});

it("单个账号留空模型时仅该账号继承默认", () => {
  const second = { ...created, local_group_ids: [4] };
  const models = {
    [onboardingAccountPreviewID(created)]: ["gpt-5.2"],
    [onboardingAccountPreviewID(second)]: [],
  };
  expect(withOnboardingProbeModels([created, second], models)).toEqual([
    { ...created, test_models: ["gpt-5.2"] },
    { ...second, test_models: [] },
  ]);
});

it("账号与已确认模型配置不匹配时拒绝提交", () => {
  expect(() => withOnboardingProbeModels([created], {})).toThrow("重新预览");
});
